src/librustc_typeck/check/coercion.rs

```diff
@@ -162,6 +163,26 @@ impl<'f, 'gcx, 'tcx> Coerce<'f, 'gcx, 'tcx> {
         }
 
         if a.is_never() {
+            // Subtle: If we are coercing from `!` to `?T`, where `?T` is an unbound
+            // type variable, we *typically* want `?T` to fallback to `!` if not
+            // otherwise constrained. An example where this arises:
+            //
+            //     let _: Option<?T> = Some({ return; });
+            //
+            // here, we would coerce from `!` to `?T`. However, we do
+            // NOT want to do this in LUB mode, since we do not want
+            // this match to get the same treatment:
+            //
+            //     let _: Option<?T> = match x {
+            //         Ok(_) => Default::default(),
```

Comment:
Ah, good catch! I am not entirely happy with that bit of code, but it seemed like the most expedient way to get the correct behavior. I had hoped to revisit it after introducing subtyping obligations (`?T <: ?U`). In any case, I'll add some examples with the match arms reversed and see what happens...